File: symfony/src/Service/Tournaments/TournamentPrizeService.php
<?php

namespace App\Service\Tournaments;

use App\Entity\Tournaments\TournamentPrize;
use App\Repository\Tournaments\TournamentPrizeRepository;

class TournamentPrizeService
{
    public function __construct(
        private TournamentPrizeRepository $repository,
    ) {}

    public function create(array $data): TournamentPrize
    {
        throw new \LogicException('Not implemented');
    }

    public function update(TournamentPrize $entity, array $data): TournamentPrize
    {
        throw new \LogicException('Not implemented');
    }

    public function appliesToPlacement(int $id, $placement): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TournamentPrize not found: ' . $id);
        $result = $entity->appliesToPlacement($placement);
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function awardToPlayer(int $id, $playerId): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('TournamentPrize not found: ' . $id);
        $entity->awardToPlayer($playerId);
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/tournaments/tournament_prize_service.rb
module Tournaments
  class TournamentPrizeService
    def initialize(repository = TournamentPrize)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def applies_to_placement(id)
      instance = TournamentPrize.find(id)
      result = instance.applies_to_placement(placement)
      instance.save!
      result
    end

    def award_to_player(id, player_id)
      instance = TournamentPrize.find(id)
      instance.award_to_player(player_id)
      instance.save!
    end
  end
end
